sw/rust/demo/led_hal/src/main.rs

```diff
@@ -0,0 +1,84 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#![no_main]
+#![no_std]
+
+extern crate panic_halt as _;
+
+use core::fmt::Write;
+use embedded_hal::blocking::delay::DelayMs;
+use riscv::delay::McycleDelay;
+use riscv_rt::entry;
+
+use embedded_hal;
+use embedded_hal::digital::v2::{InputPin, OutputPin};
+use hal::{pac, ErasePin, GpioExt};
+use ibex_demo_system_hal as hal;
+
+const CPU_CLOCK_HZ: u32 = 50_000_000;
+
+struct BtnLedMap<'a> {
+    btn: &'a dyn InputPin<Error = hal::utils::Error>,
+    led: hal::gpio::DynPin,
+    prev_val: bool,
+}
+
+impl<'a> BtnLedMap<'a> {
+    fn new(
+        btn: &'a dyn InputPin<Error = hal::utils::Error>,
+        led: hal::gpio::DynPin,
+        prev_val: bool,
+    ) -> Self {
+        Self { btn, led, prev_val }
+    }
+}
+
+#[entry]
+fn main() -> ! {
+    let mut delay = McycleDelay::new(CPU_CLOCK_HZ);
+    let p = pac::Peripherals::take().unwrap();
+
+    let mut uart = hal::serial::Serial::new(p.UART0);
+
+    let pins = p.GPIOA.pins();
+    let btn0 = pins.pin0.into_input();
+    let btn1 = pins.pin1.into_input();
+    let btn2 = pins.pin2.into_input();
+    let btn3 = pins.pin3.into_input();
+
+    // Dynamic pins are convenient because they have the same type, but theres a run time cost.
+    let led0 = pins.pin4.into_dynamic().into_output().unwrap();
+    let led1 = pins.pin5.into_dynamic().into_output().unwrap();
+    let led2 = pins.pin6.into_dynamic().into_output().unwrap();
+    let led3 = pins.pin7.into_dynamic().into_output().unwrap();
+
+    let mut map = [
+        BtnLedMap::new(&btn0, led0, false),
+        BtnLedMap::new(&btn1, led1, false),
+        BtnLedMap::new(&btn2, led2, false),
+        BtnLedMap::new(&btn3, led3, false),
+    ];
+
+    let _ = writeln!(uart, "Hello Rusty Ibex System!!");
+    let _ = writeln!(uart, "Press a button (BTN0-4) and watch the LEDs");
+
+    loop {
+        for node in map.iter_mut() {
+            if let Ok(in_val) = node.btn.is_high() {
```

Comment:
Is there a single error type in HAL? It might worth making the return type `Result<(), Error>` and use `?` here.

Reply:
the `#[entry]` macro requires the main to have that signature, so I swapped to a `unwrap` instead.